In [261]:
#import all the libaries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [262]:
#read the dataset 
data=pd.read_excel("Flight_Fare.xlsx")
data

,Airline,Date_of_Journey,Source,Destination,Route,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info,Price
0,IndiGo,24/03/2019,Banglore,New Delhi,BLR → DEL,22:20,01:10 22 Mar,2h 50m,non-stop,No info,3897
1,Air India,1/05/2019,Kolkata,Banglore,CCU → IXR → BBI → BLR,05:50,13:15,7h 25m,2 stops,No info,7662
2,Jet Airways,9/06/2019,Delhi,Cochin,DEL → LKO → BOM → COK,09:25,04:25 10 Jun,19h,2 stops,No info,13882
3,IndiGo,12/05/2019,Kolkata,Banglore,CCU → NAG → BLR,18:05,23:30,5h 25m,1 stop,No info,6218
4,IndiGo,01/03/2019,Banglore,New Delhi,BLR → NAG → DEL,16:50,21:35,4h 45m,1 stop,No info,13302
...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,9/04/2019,Kolkata,Banglore,CCU → BLR,19:55,22:25,2h 30m,non-stop,No info,4107
10679,Air India,27/04/2019,Kolkata,Banglore,CCU → BLR,20:45,23:20,2h 35m,non-stop,No info,4145
10680,Jet Airways,27/04/2019,Banglore,Delhi,BLR → DEL,08:20,11:20,3h,non-stop,No info,7229
10681,Vistara,01/03/2019,Banglore,New Delhi,BLR → DEL,11:30,14:10,2h 40m,non-stop,No info,12648


In [263]:
#copy original dataset
df=data.copy()
df

,Airline,Date_of_Journey,Source,Destination,Route,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info,Price
0,IndiGo,24/03/2019,Banglore,New Delhi,BLR → DEL,22:20,01:10 22 Mar,2h 50m,non-stop,No info,3897
1,Air India,1/05/2019,Kolkata,Banglore,CCU → IXR → BBI → BLR,05:50,13:15,7h 25m,2 stops,No info,7662
2,Jet Airways,9/06/2019,Delhi,Cochin,DEL → LKO → BOM → COK,09:25,04:25 10 Jun,19h,2 stops,No info,13882
3,IndiGo,12/05/2019,Kolkata,Banglore,CCU → NAG → BLR,18:05,23:30,5h 25m,1 stop,No info,6218
4,IndiGo,01/03/2019,Banglore,New Delhi,BLR → NAG → DEL,16:50,21:35,4h 45m,1 stop,No info,13302
...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,9/04/2019,Kolkata,Banglore,CCU → BLR,19:55,22:25,2h 30m,non-stop,No info,4107
10679,Air India,27/04/2019,Kolkata,Banglore,CCU → BLR,20:45,23:20,2h 35m,non-stop,No info,4145
10680,Jet Airways,27/04/2019,Banglore,Delhi,BLR → DEL,08:20,11:20,3h,non-stop,No info,7229
10681,Vistara,01/03/2019,Banglore,New Delhi,BLR → DEL,11:30,14:10,2h 40m,non-stop,No info,12648


In [264]:
#check first 5 raws 
print(df.head())


       Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302  


In [265]:
#check last 5 raws
print(df.tail())

           Airline Date_of_Journey    Source Destination  \
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time Arrival_Time Duration Total_Stops  \
10678              CCU → BLR    19:55        22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45        23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20        11:20       3h    non-stop   
10681              BLR → DEL    11:30        14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55        19:15   8h 20m     2 stops   

      Additional_Info  Price  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  


In [266]:
#check the no.of raws and columns
df.shape

(10683, 11)

In [267]:
'''Statistical summary was generated using df.describe(),
to understand the distribution, central tendency, and spread of numerical features.'''
df.describe()

,Price
count,10683.000000
mean,9087.064121
std,4611.359167
min,1759.000000
25%,5277.000000
50%,8372.000000
75%,12373.000000
max,79512.000000


In [268]:
'''Categorical feature summary was analyzed 
to identify the most frequent categories and understand feature distribution'''
df.describe(include='O')

,Airline,Date_of_Journey,Source,Destination,Route,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info
count,10683,10683,10683,10683,10682,10683,10683,10683,10682,10683
unique,12,44,5,6,128,222,1343,368,5,10
top,Jet Airways,18/05/2019,Delhi,Cochin,DEL → BOM → COK,18:55,19:00,2h 50m,1 stop,No info
freq,3849,504,4537,4537,2376,233,423,550,5625,8345


In [269]:
'''df.info() was used to inspect column details, 
identify data types, and check for missing values before preprocessing.'''
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [270]:
'''df.isnull().sum() was used to identify missing values in the dataset 
so appropriate preprocessing techniques could be applied.'''
df.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

In [271]:
'''Missing values in Total_Stops and Route were handled using mode ,
as these are categorical features and the most frequent category helps consistency.'''
df['Total_Stops']=df['Total_Stops'].fillna(df['Total_Stops'].mode()[0])
df['Route']=df['Route'].fillna(df['Route'].mode()[0])

In [272]:
'''After handling missing values using mode, 
df.isnull().sum() was used again to verify that the dataset no longer contained null values.'''
df.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              0
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        0
Additional_Info    0
Price              0
dtype: int64

In [273]:
'''df.duplicated().sum() was used to identify duplicate records in the dataset
to ensure data quality and avoid biased model training.'''
df.duplicated().sum()

np.int64(220)

In [274]:
'''Duplicate records were inspected using df[df.duplicated()].head() 
to verify repeated entries before removing them from the dataset'''
df[df.duplicated()].head()

,Airline,Date_of_Journey,Source,Destination,Route,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info,Price
683,Jet Airways,1/06/2019,Delhi,Cochin,DEL → NAG → BOM → COK,14:35,04:25 02 Jun,13h 50m,2 stops,No info,13376
1061,Air India,21/05/2019,Delhi,Cochin,DEL → GOI → BOM → COK,22:00,19:15 22 May,21h 15m,2 stops,No info,10231
1348,Air India,18/05/2019,Delhi,Cochin,DEL → HYD → BOM → COK,17:15,19:15 19 May,26h,2 stops,No info,12392
1418,Jet Airways,6/06/2019,Delhi,Cochin,DEL → JAI → BOM → COK,05:30,04:25 07 Jun,22h 55m,2 stops,In-flight meal not included,10368
1674,IndiGo,24/03/2019,Banglore,New Delhi,BLR → DEL,18:25,21:20,2h 55m,non-stop,No info,7303


In [276]:
'''Duplicate rows were removed using drop_duplicates()
to improve data quality, avoid redundant information, and prevent biased model training.'''
df.drop_duplicates(inplace=True)

In [203]:
'''After removing duplicate records, df.duplicated().sum()
was used again to verify that no duplicate rows remained in the dataset.'''
df.duplicated().sum()

np.int64(0)

In [277]:
'''Machine learning models cannot understand date values in string/object format. 
They work only with numerical data, so Date_of_Journey was converted into datetime format using pd.to_datetime()
and useful numerical features like day, month, and year'''
df['day']=pd.to_datetime(df['Date_of_Journey'],format="%d/%m/%Y").dt.day
df['month']=pd.to_datetime(df['Date_of_Journey'],format="%d/%m/%Y").dt.month
df['year']=pd.to_datetime(df['Date_of_Journey'],format="%d/%m/%Y").dt.year
print(df)

           Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun

In [278]:
'''After extracting day, month, and year features, the original Date_of_Journey column was dropped 
because machine learning models require numerical features and the original date column became redundant'''
df.drop("Date_of_Journey", axis=1, inplace=True)
df

,Airline,Source,Destination,Route,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info,Price,day,month,year
0,IndiGo,Banglore,New Delhi,BLR → DEL,22:20,01:10 22 Mar,2h 50m,non-stop,No info,3897,24,3,2019
1,Air India,Kolkata,Banglore,CCU → IXR → BBI → BLR,05:50,13:15,7h 25m,2 stops,No info,7662,1,5,2019
2,Jet Airways,Delhi,Cochin,DEL → LKO → BOM → COK,09:25,04:25 10 Jun,19h,2 stops,No info,13882,9,6,2019
3,IndiGo,Kolkata,Banglore,CCU → NAG → BLR,18:05,23:30,5h 25m,1 stop,No info,6218,12,5,2019
4,IndiGo,Banglore,New Delhi,BLR → NAG → DEL,16:50,21:35,4h 45m,1 stop,No info,13302,1,3,2019
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,Kolkata,Banglore,CCU → BLR,19:55,22:25,2h 30m,non-stop,No info,4107,9,4,2019
10679,Air India,Kolkata,Banglore,CCU → BLR,20:45,23:20,2h 35m,non-stop,No info,4145,27,4,2019
10680,Jet Airways,Banglore,Delhi,BLR → DEL,08:20,11:20,3h,non-stop,No info,7229,27,4,2019
10681,Vistara,Banglore,New Delhi,BLR → DEL,11:30,14:10,2h 40m,non-stop,No info,12648,1,3,2019


In [279]:
#The Route column was dropped because similar information was already represented by Total_Stops 
df.drop(columns="Route",axis=1,inplace=True)

In [280]:
df

,Airline,Source,Destination,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info,Price,day,month,year
0,IndiGo,Banglore,New Delhi,22:20,01:10 22 Mar,2h 50m,non-stop,No info,3897,24,3,2019
1,Air India,Kolkata,Banglore,05:50,13:15,7h 25m,2 stops,No info,7662,1,5,2019
2,Jet Airways,Delhi,Cochin,09:25,04:25 10 Jun,19h,2 stops,No info,13882,9,6,2019
3,IndiGo,Kolkata,Banglore,18:05,23:30,5h 25m,1 stop,No info,6218,12,5,2019
4,IndiGo,Banglore,New Delhi,16:50,21:35,4h 45m,1 stop,No info,13302,1,3,2019
...,...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,Kolkata,Banglore,19:55,22:25,2h 30m,non-stop,No info,4107,9,4,2019
10679,Air India,Kolkata,Banglore,20:45,23:20,2h 35m,non-stop,No info,4145,27,4,2019
10680,Jet Airways,Banglore,Delhi,08:20,11:20,3h,non-stop,No info,7229,27,4,2019
10681,Vistara,Banglore,New Delhi,11:30,14:10,2h 40m,non-stop,No info,12648,1,3,2019


In [281]:
'''Dep_Time was converted into datetime format using pd.to_datetime() 
because machine learning models require numerical input. 
Hour and minute features were extracted to capture the effect of departure timing'''
df['hours']=pd.to_datetime(df['Dep_Time'],format="%H:%M").dt.hour
df['minutes']=pd.to_datetime(df['Dep_Time'],format="%H:%M").dt.minute

In [282]:
df

,Airline,Source,Destination,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info,Price,day,month,year,hours,minutes
0,IndiGo,Banglore,New Delhi,22:20,01:10 22 Mar,2h 50m,non-stop,No info,3897,24,3,2019,22,20
1,Air India,Kolkata,Banglore,05:50,13:15,7h 25m,2 stops,No info,7662,1,5,2019,5,50
2,Jet Airways,Delhi,Cochin,09:25,04:25 10 Jun,19h,2 stops,No info,13882,9,6,2019,9,25
3,IndiGo,Kolkata,Banglore,18:05,23:30,5h 25m,1 stop,No info,6218,12,5,2019,18,5
4,IndiGo,Banglore,New Delhi,16:50,21:35,4h 45m,1 stop,No info,13302,1,3,2019,16,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,Kolkata,Banglore,19:55,22:25,2h 30m,non-stop,No info,4107,9,4,2019,19,55
10679,Air India,Kolkata,Banglore,20:45,23:20,2h 35m,non-stop,No info,4145,27,4,2019,20,45
10680,Jet Airways,Banglore,Delhi,08:20,11:20,3h,non-stop,No info,7229,27,4,2019,8,20
10681,Vistara,Banglore,New Delhi,11:30,14:10,2h 40m,non-stop,No info,12648,1,3,2019,11,30


In [283]:
'''After extracting hour and minute features from Dep_Time, the original column was dropped 
because machine learning models require numerical data and the original time column became redundant.'''
df.drop(columns='Dep_Time',inplace=True)

In [284]:
#check the dataset
df

,Airline,Source,Destination,Arrival_Time,Duration,Total_Stops,Additional_Info,Price,day,month,year,hours,minutes
0,IndiGo,Banglore,New Delhi,01:10 22 Mar,2h 50m,non-stop,No info,3897,24,3,2019,22,20
1,Air India,Kolkata,Banglore,13:15,7h 25m,2 stops,No info,7662,1,5,2019,5,50
2,Jet Airways,Delhi,Cochin,04:25 10 Jun,19h,2 stops,No info,13882,9,6,2019,9,25
3,IndiGo,Kolkata,Banglore,23:30,5h 25m,1 stop,No info,6218,12,5,2019,18,5
4,IndiGo,Banglore,New Delhi,21:35,4h 45m,1 stop,No info,13302,1,3,2019,16,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,Kolkata,Banglore,22:25,2h 30m,non-stop,No info,4107,9,4,2019,19,55
10679,Air India,Kolkata,Banglore,23:20,2h 35m,non-stop,No info,4145,27,4,2019,20,45
10680,Jet Airways,Banglore,Delhi,11:20,3h,non-stop,No info,7229,27,4,2019,8,20
10681,Vistara,Banglore,New Delhi,14:10,2h 40m,non-stop,No info,12648,1,3,2019,11,30


In [285]:
'''Arrival_Time was converted into datetime format to extract numerical features such as arrival hour and minute. 
errors='coerce' was used to handle inconsistent time formats(here we have hh:mm and hh:mm month formats) without interrupting preprocessing.'''
df["Arrival_Time"] = pd.to_datetime(df["Arrival_Time"], errors="coerce")
df["Arrival_hour"] = df["Arrival_Time"].dt.hour
df["Arrival_min"] = df["Arrival_Time"].dt.minute

C:\Users\PRABHITH VK\AppData\Local\Temp\ipykernel_10700\2510145163.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df["Arrival_Time"] = pd.to_datetime(df["Arrival_Time"], errors="coerce")


In [286]:
df

,Airline,Source,Destination,Arrival_Time,Duration,Total_Stops,Additional_Info,Price,day,month,year,hours,minutes,Arrival_hour,Arrival_min
0,IndiGo,Banglore,New Delhi,2026-03-22 01:10:00,2h 50m,non-stop,No info,3897,24,3,2019,22,20,1,10
1,Air India,Kolkata,Banglore,2026-05-28 13:15:00,7h 25m,2 stops,No info,7662,1,5,2019,5,50,13,15
2,Jet Airways,Delhi,Cochin,2026-06-10 04:25:00,19h,2 stops,No info,13882,9,6,2019,9,25,4,25
3,IndiGo,Kolkata,Banglore,2026-05-28 23:30:00,5h 25m,1 stop,No info,6218,12,5,2019,18,5,23,30
4,IndiGo,Banglore,New Delhi,2026-05-28 21:35:00,4h 45m,1 stop,No info,13302,1,3,2019,16,50,21,35
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,Kolkata,Banglore,2026-05-28 22:25:00,2h 30m,non-stop,No info,4107,9,4,2019,19,55,22,25
10679,Air India,Kolkata,Banglore,2026-05-28 23:20:00,2h 35m,non-stop,No info,4145,27,4,2019,20,45,23,20
10680,Jet Airways,Banglore,Delhi,2026-05-28 11:20:00,3h,non-stop,No info,7229,27,4,2019,8,20,11,20
10681,Vistara,Banglore,New Delhi,2026-05-28 14:10:00,2h 40m,non-stop,No info,12648,1,3,2019,11,30,14,10


In [287]:
'''The columns were renamed to Dept_hours and Dept_minutes for better readability 
and clearer understanding of the extracted departure time features'''
df.rename(columns={"hours": "Dept_hours","minutes": "Dept_minutes"}, inplace=True)

In [288]:
df

,Airline,Source,Destination,Arrival_Time,Duration,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,Arrival_min
0,IndiGo,Banglore,New Delhi,2026-03-22 01:10:00,2h 50m,non-stop,No info,3897,24,3,2019,22,20,1,10
1,Air India,Kolkata,Banglore,2026-05-28 13:15:00,7h 25m,2 stops,No info,7662,1,5,2019,5,50,13,15
2,Jet Airways,Delhi,Cochin,2026-06-10 04:25:00,19h,2 stops,No info,13882,9,6,2019,9,25,4,25
3,IndiGo,Kolkata,Banglore,2026-05-28 23:30:00,5h 25m,1 stop,No info,6218,12,5,2019,18,5,23,30
4,IndiGo,Banglore,New Delhi,2026-05-28 21:35:00,4h 45m,1 stop,No info,13302,1,3,2019,16,50,21,35
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,Kolkata,Banglore,2026-05-28 22:25:00,2h 30m,non-stop,No info,4107,9,4,2019,19,55,22,25
10679,Air India,Kolkata,Banglore,2026-05-28 23:20:00,2h 35m,non-stop,No info,4145,27,4,2019,20,45,23,20
10680,Jet Airways,Banglore,Delhi,2026-05-28 11:20:00,3h,non-stop,No info,7229,27,4,2019,8,20,11,20
10681,Vistara,Banglore,New Delhi,2026-05-28 14:10:00,2h 40m,non-stop,No info,12648,1,3,2019,11,30,14,10


In [289]:
'''After extracting numerical features such as Arrival_hour and Arrival_min, 
the original Arrival_Time column was dropped because it became redundant and machine learning models require numerical input features'''
df.drop('Arrival_Time',axis=1,inplace=True)

In [290]:
df

,Airline,Source,Destination,Duration,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,Arrival_min
0,IndiGo,Banglore,New Delhi,2h 50m,non-stop,No info,3897,24,3,2019,22,20,1,10
1,Air India,Kolkata,Banglore,7h 25m,2 stops,No info,7662,1,5,2019,5,50,13,15
2,Jet Airways,Delhi,Cochin,19h,2 stops,No info,13882,9,6,2019,9,25,4,25
3,IndiGo,Kolkata,Banglore,5h 25m,1 stop,No info,6218,12,5,2019,18,5,23,30
4,IndiGo,Banglore,New Delhi,4h 45m,1 stop,No info,13302,1,3,2019,16,50,21,35
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,Air Asia,Kolkata,Banglore,2h 30m,non-stop,No info,4107,9,4,2019,19,55,22,25
10679,Air India,Kolkata,Banglore,2h 35m,non-stop,No info,4145,27,4,2019,20,45,23,20
10680,Jet Airways,Banglore,Delhi,3h,non-stop,No info,7229,27,4,2019,8,20,11,20
10681,Vistara,Banglore,New Delhi,2h 40m,non-stop,No info,12648,1,3,2019,11,30,14,10


In [291]:
'''One-Hot Encoding was used for the Airline column because airline names are categorical data without any order or ranking.
Label Encoding was not used because it assigns numbers to categories, 
and the model may misunderstand them as ordered values. 
One-Hot Encoding converts each airline into separate 0 and 1 columns, making it suitable for machine learning models.'''
df = pd.get_dummies(df, columns=["Airline"], dtype=int)

In [292]:
df

,Source,Destination,Duration,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,...,Airline_GoAir,Airline_IndiGo,Airline_Jet Airways,Airline_Jet Airways Business,Airline_Multiple carriers,Airline_Multiple carriers Premium economy,Airline_SpiceJet,Airline_Trujet,Airline_Vistara,Airline_Vistara Premium economy
0,Banglore,New Delhi,2h 50m,non-stop,No info,3897,24,3,2019,22,...,0,1,0,0,0,0,0,0,0,0
1,Kolkata,Banglore,7h 25m,2 stops,No info,7662,1,5,2019,5,...,0,0,0,0,0,0,0,0,0,0
2,Delhi,Cochin,19h,2 stops,No info,13882,9,6,2019,9,...,0,0,1,0,0,0,0,0,0,0
3,Kolkata,Banglore,5h 25m,1 stop,No info,6218,12,5,2019,18,...,0,1,0,0,0,0,0,0,0,0
4,Banglore,New Delhi,4h 45m,1 stop,No info,13302,1,3,2019,16,...,0,1,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,Kolkata,Banglore,2h 30m,non-stop,No info,4107,9,4,2019,19,...,0,0,0,0,0,0,0,0,0,0
10679,Kolkata,Banglore,2h 35m,non-stop,No info,4145,27,4,2019,20,...,0,0,0,0,0,0,0,0,0,0
10680,Banglore,Delhi,3h,non-stop,No info,7229,27,4,2019,8,...,0,0,1,0,0,0,0,0,0,0
10681,Banglore,New Delhi,2h 40m,non-stop,No info,12648,1,3,2019,11,...,0,0,0,0,0,0,0,0,1,0


In [293]:
'''One-Hot Encoding was applied to Source and Destination because they are categorical features without any ordinal relationship. 
It converts categories into binary (0 and 1) columns, making them suitable for machine learning models'''
df = pd.get_dummies(df, columns=["Source", "Destination"], dtype=int)

In [294]:
df

,Duration,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,...,Source_Chennai,Source_Delhi,Source_Kolkata,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi
0,2h 50m,non-stop,No info,3897,24,3,2019,22,20,1,...,0,0,0,0,0,0,0,0,0,1
1,7h 25m,2 stops,No info,7662,1,5,2019,5,50,13,...,0,0,1,0,1,0,0,0,0,0
2,19h,2 stops,No info,13882,9,6,2019,9,25,4,...,0,1,0,0,0,1,0,0,0,0
3,5h 25m,1 stop,No info,6218,12,5,2019,18,5,23,...,0,0,1,0,1,0,0,0,0,0
4,4h 45m,1 stop,No info,13302,1,3,2019,16,50,21,...,0,0,0,0,0,0,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,2h 30m,non-stop,No info,4107,9,4,2019,19,55,22,...,0,0,1,0,1,0,0,0,0,0
10679,2h 35m,non-stop,No info,4145,27,4,2019,20,45,23,...,0,0,1,0,1,0,0,0,0,0
10680,3h,non-stop,No info,7229,27,4,2019,8,20,11,...,0,0,0,0,0,0,1,0,0,0
10681,2h 40m,non-stop,No info,12648,1,3,2019,11,30,14,...,0,0,0,0,0,0,0,0,0,1


In [295]:
df

,Duration,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,...,Source_Chennai,Source_Delhi,Source_Kolkata,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi
0,2h 50m,non-stop,No info,3897,24,3,2019,22,20,1,...,0,0,0,0,0,0,0,0,0,1
1,7h 25m,2 stops,No info,7662,1,5,2019,5,50,13,...,0,0,1,0,1,0,0,0,0,0
2,19h,2 stops,No info,13882,9,6,2019,9,25,4,...,0,1,0,0,0,1,0,0,0,0
3,5h 25m,1 stop,No info,6218,12,5,2019,18,5,23,...,0,0,1,0,1,0,0,0,0,0
4,4h 45m,1 stop,No info,13302,1,3,2019,16,50,21,...,0,0,0,0,0,0,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,2h 30m,non-stop,No info,4107,9,4,2019,19,55,22,...,0,0,1,0,1,0,0,0,0,0
10679,2h 35m,non-stop,No info,4145,27,4,2019,20,45,23,...,0,0,1,0,1,0,0,0,0,0
10680,3h,non-stop,No info,7229,27,4,2019,8,20,11,...,0,0,0,0,0,0,1,0,0,0
10681,2h 40m,non-stop,No info,12648,1,3,2019,11,30,14,...,0,0,0,0,0,0,0,0,0,1


In [296]:
'''Duration was not converted using pd.to_datetime() because it represents travel duration 
rather than actual clock time. The values were manually converted into total minutes
to create meaningful numerical features for machine learning models.'''
df['Duration'] = df['Duration'].str.replace('h', '*60 +').str.replace('m', '*1')
df['Duration_minutes'] = df['Duration'].apply(eval)

SyntaxError: invalid syntax (<string>, line 1)

In [297]:
df

,Duration,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,...,Source_Chennai,Source_Delhi,Source_Kolkata,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi
0,2*60 + 50*1,non-stop,No info,3897,24,3,2019,22,20,1,...,0,0,0,0,0,0,0,0,0,1
1,7*60 + 25*1,2 stops,No info,7662,1,5,2019,5,50,13,...,0,0,1,0,1,0,0,0,0,0
2,19*60 +,2 stops,No info,13882,9,6,2019,9,25,4,...,0,1,0,0,0,1,0,0,0,0
3,5*60 + 25*1,1 stop,No info,6218,12,5,2019,18,5,23,...,0,0,1,0,1,0,0,0,0,0
4,4*60 + 45*1,1 stop,No info,13302,1,3,2019,16,50,21,...,0,0,0,0,0,0,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,2*60 + 30*1,non-stop,No info,4107,9,4,2019,19,55,22,...,0,0,1,0,1,0,0,0,0,0
10679,2*60 + 35*1,non-stop,No info,4145,27,4,2019,20,45,23,...,0,0,1,0,1,0,0,0,0,0
10680,3*60 +,non-stop,No info,7229,27,4,2019,8,20,11,...,0,0,0,0,0,0,1,0,0,0
10681,2*60 + 40*1,non-stop,No info,12648,1,3,2019,11,30,14,...,0,0,0,0,0,0,0,0,0,1


In [298]:
'''The modified Duration values were converted back into readable hour-minute format
using string replacement functions for proper feature extraction.'''
df['Duration'] = df['Duration'].str.replace('*60 +', 'h ', regex=False)

df['Duration'] = df['Duration'].str.replace('*1', 'm', regex=False)

In [299]:
df

,Duration,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,...,Source_Chennai,Source_Delhi,Source_Kolkata,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi
0,2h 50m,non-stop,No info,3897,24,3,2019,22,20,1,...,0,0,0,0,0,0,0,0,0,1
1,7h 25m,2 stops,No info,7662,1,5,2019,5,50,13,...,0,0,1,0,1,0,0,0,0,0
2,19h,2 stops,No info,13882,9,6,2019,9,25,4,...,0,1,0,0,0,1,0,0,0,0
3,5h 25m,1 stop,No info,6218,12,5,2019,18,5,23,...,0,0,1,0,1,0,0,0,0,0
4,4h 45m,1 stop,No info,13302,1,3,2019,16,50,21,...,0,0,0,0,0,0,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,2h 30m,non-stop,No info,4107,9,4,2019,19,55,22,...,0,0,1,0,1,0,0,0,0,0
10679,2h 35m,non-stop,No info,4145,27,4,2019,20,45,23,...,0,0,1,0,1,0,0,0,0,0
10680,3h,non-stop,No info,7229,27,4,2019,8,20,11,...,0,0,0,0,0,0,1,0,0,0
10681,2h 40m,non-stop,No info,12648,1,3,2019,11,30,14,...,0,0,0,0,0,0,0,0,0,1


In [300]:
'''The Duration column contained mixed time formats, so hours and minutes were extracted separately using regular expressions.
The extracted values were converted into numerical format and combined into total minutes to create a consistent feature suitable
for machine learning models.'''
# extract hours
df['dur_hour'] = df['Duration'].str.extract('(\d+)h')

# extract minutes
df['dur_min'] = df['Duration'].str.extract('(\d+)m')

# convert to numeric
df['dur_hour'] = pd.to_numeric(df['dur_hour'], errors='coerce').fillna(0).astype(int)

df['dur_min'] = pd.to_numeric(df['dur_min'], errors='coerce').fillna(0).astype(int)

# total duration in minutes
df['Duration_total_mins'] = df['dur_hour'] * 60 + df['dur_min']

In [301]:
df

,Duration,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,...,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi,dur_hour,dur_min,Duration_total_mins
0,2h 50m,non-stop,No info,3897,24,3,2019,22,20,1,...,0,0,0,0,0,0,1,2,50,170
1,7h 25m,2 stops,No info,7662,1,5,2019,5,50,13,...,0,1,0,0,0,0,0,7,25,445
2,19h,2 stops,No info,13882,9,6,2019,9,25,4,...,0,0,1,0,0,0,0,19,0,1140
3,5h 25m,1 stop,No info,6218,12,5,2019,18,5,23,...,0,1,0,0,0,0,0,5,25,325
4,4h 45m,1 stop,No info,13302,1,3,2019,16,50,21,...,0,0,0,0,0,0,1,4,45,285
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,2h 30m,non-stop,No info,4107,9,4,2019,19,55,22,...,0,1,0,0,0,0,0,2,30,150
10679,2h 35m,non-stop,No info,4145,27,4,2019,20,45,23,...,0,1,0,0,0,0,0,2,35,155
10680,3h,non-stop,No info,7229,27,4,2019,8,20,11,...,0,0,0,1,0,0,0,3,0,180
10681,2h 40m,non-stop,No info,12648,1,3,2019,11,30,14,...,0,0,0,0,0,0,1,2,40,160


In [302]:
'''After extracting useful numerical features from the Duration column and converting them into total minutes, 
the original Duration column was dropped because it became redundant and machine learning models require numerical input features'''
df.drop('Duration', axis=1, inplace=True)

In [303]:
df

,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,Arrival_min,...,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi,dur_hour,dur_min,Duration_total_mins
0,non-stop,No info,3897,24,3,2019,22,20,1,10,...,0,0,0,0,0,0,1,2,50,170
1,2 stops,No info,7662,1,5,2019,5,50,13,15,...,0,1,0,0,0,0,0,7,25,445
2,2 stops,No info,13882,9,6,2019,9,25,4,25,...,0,0,1,0,0,0,0,19,0,1140
3,1 stop,No info,6218,12,5,2019,18,5,23,30,...,0,1,0,0,0,0,0,5,25,325
4,1 stop,No info,13302,1,3,2019,16,50,21,35,...,0,0,0,0,0,0,1,4,45,285
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,non-stop,No info,4107,9,4,2019,19,55,22,25,...,0,1,0,0,0,0,0,2,30,150
10679,non-stop,No info,4145,27,4,2019,20,45,23,20,...,0,1,0,0,0,0,0,2,35,155
10680,non-stop,No info,7229,27,4,2019,8,20,11,20,...,0,0,0,1,0,0,0,3,0,180
10681,non-stop,No info,12648,1,3,2019,11,30,14,10,...,0,0,0,0,0,0,1,2,40,160


In [304]:
'''Total_Stops was converted into numerical values using mapping because it is an ordinal categorical feature 
with a meaningful order between categories.'''
df['Total_Stops'] = df['Total_Stops'].map({
    'non-stop': 0,
    '1 stop': 1,
    '2 stops': 2,
    '3 stops': 3,
    '4 stops': 4
}).astype(int)

In [305]:
'''While converting Total_Stops into numerical values using mapping, 
the missing values (NaN) were not included in the mapping step. 
These missing values were later handled separately using fillna() to ensure there are no null values in the dataset 
and to make the feature suitable for machine learning models.'''
df['Total_Stops'] = df['Total_Stops'].fillna(5).astype(int)

In [306]:
df['Total_Stops'].unique()

array([0, 2, 1, 3, 4])

In [307]:
df

,Total_Stops,Additional_Info,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,Arrival_min,...,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi,dur_hour,dur_min,Duration_total_mins
0,0,No info,3897,24,3,2019,22,20,1,10,...,0,0,0,0,0,0,1,2,50,170
1,2,No info,7662,1,5,2019,5,50,13,15,...,0,1,0,0,0,0,0,7,25,445
2,2,No info,13882,9,6,2019,9,25,4,25,...,0,0,1,0,0,0,0,19,0,1140
3,1,No info,6218,12,5,2019,18,5,23,30,...,0,1,0,0,0,0,0,5,25,325
4,1,No info,13302,1,3,2019,16,50,21,35,...,0,0,0,0,0,0,1,4,45,285
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,0,No info,4107,9,4,2019,19,55,22,25,...,0,1,0,0,0,0,0,2,30,150
10679,0,No info,4145,27,4,2019,20,45,23,20,...,0,1,0,0,0,0,0,2,35,155
10680,0,No info,7229,27,4,2019,8,20,11,20,...,0,0,0,1,0,0,0,3,0,180
10681,0,No info,12648,1,3,2019,11,30,14,10,...,0,0,0,0,0,0,1,2,40,160


In [308]:
df['Additional_Info'].value_counts()

Additional_Info
No info                         8183
In-flight meal not included     1926
No check-in baggage included     318
1 Long layover                    19
Change airports                    7
Business class                     4
No Info                            3
1 Short layover                    1
Red-eye flight                     1
2 Long layover                     1
Name: count, dtype: int64

In [309]:
'''The Additional_Info column was dropped because it is highly imbalanced, with most values being ‘No info’. 
The remaining categories have very low frequency, making the feature less useful for machine learning and likely to introduce noise.'''
df.drop('Additional_Info', axis=1, inplace=True)

In [310]:
df

,Total_Stops,Price,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,Arrival_min,Airline_Air Asia,...,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi,dur_hour,dur_min,Duration_total_mins
0,0,3897,24,3,2019,22,20,1,10,0,...,0,0,0,0,0,0,1,2,50,170
1,2,7662,1,5,2019,5,50,13,15,0,...,0,1,0,0,0,0,0,7,25,445
2,2,13882,9,6,2019,9,25,4,25,0,...,0,0,1,0,0,0,0,19,0,1140
3,1,6218,12,5,2019,18,5,23,30,0,...,0,1,0,0,0,0,0,5,25,325
4,1,13302,1,3,2019,16,50,21,35,0,...,0,0,0,0,0,0,1,4,45,285
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,0,4107,9,4,2019,19,55,22,25,1,...,0,1,0,0,0,0,0,2,30,150
10679,0,4145,27,4,2019,20,45,23,20,0,...,0,1,0,0,0,0,0,2,35,155
10680,0,7229,27,4,2019,8,20,11,20,0,...,0,0,0,1,0,0,0,3,0,180
10681,0,12648,1,3,2019,11,30,14,10,0,...,0,0,0,0,0,0,1,2,40,160


In [311]:
#Separated input features (X) and target variable (y) for machine learning model building.X has all the features except price and y has only price
X = df.drop("Price", axis=1)
y = df["Price"]

In [312]:
X

,Total_Stops,day,month,year,Dept_hours,Dept_minutes,Arrival_hour,Arrival_min,Airline_Air Asia,Airline_Air India,...,Source_Mumbai,Destination_Banglore,Destination_Cochin,Destination_Delhi,Destination_Hyderabad,Destination_Kolkata,Destination_New Delhi,dur_hour,dur_min,Duration_total_mins
0,0,24,3,2019,22,20,1,10,0,0,...,0,0,0,0,0,0,1,2,50,170
1,2,1,5,2019,5,50,13,15,0,1,...,0,1,0,0,0,0,0,7,25,445
2,2,9,6,2019,9,25,4,25,0,0,...,0,0,1,0,0,0,0,19,0,1140
3,1,12,5,2019,18,5,23,30,0,0,...,0,1,0,0,0,0,0,5,25,325
4,1,1,3,2019,16,50,21,35,0,0,...,0,0,0,0,0,0,1,4,45,285
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10678,0,9,4,2019,19,55,22,25,1,0,...,0,1,0,0,0,0,0,2,30,150
10679,0,27,4,2019,20,45,23,20,0,1,...,0,1,0,0,0,0,0,2,35,155
10680,0,27,4,2019,8,20,11,20,0,0,...,0,0,0,1,0,0,0,3,0,180
10681,0,1,3,2019,11,30,14,10,0,0,...,0,0,0,0,0,0,1,2,40,160


In [313]:
y

0         3897
1         7662
2        13882
3         6218
4        13302
         ...  
10678     4107
10679     4145
10680     7229
10681    12648
10682    11753
Name: Price, Length: 10463, dtype: int64

In [314]:
#Performed train-test split to divide the dataset into training and testing sets for model training and evaluation.
#import train-test split
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [315]:
#Implemented Linear Regression model and trained it using training dataset to learn relationships between input features and flight price.
#import LinearRegression
from sklearn.linear_model import LinearRegression
#model creation
LR=LinearRegression()
#Train model
LR.fit(x_train,y_train)


LinearRegression()

In [316]:
#Generated predicted flight fare values using the trained Linear Regression model on test data.
#predict the model
y_pred=LR.predict(x_test)

In [317]:
#Evaluated Linear Regression model performance using MAE, MSE, and R² score metrics.
''''''
from sklearn.metrics import mean_squared_error,mean_absolute_error,r2_score
print("MSE",mean_squared_error(y_test,y_pred))
print("MAE",mean_absolute_error(y_test,y_pred))
print("R2",r2_score(y_test,y_pred))

MSE 8304569.255930112
MAE 1957.1738017367306
R2 0.6019798119194207


The Linear Regression model achieved an MAE of around 1957, which means the predicted flight fare differs from the actual fare by approximately ₹1957 on average. The MSE value was high because errors are squared, which heavily penalizes larger prediction errors. The model achieved an R² score of 0.60, indicating that it explains around 60% of the variance in flight prices. Since flight fare data contains nonlinear and complex patterns, Linear Regression provided moderate baseline performance, and advanced ensemble models can further improve accuracy.

In [318]:
'''StandardScaler was used instead of MinMaxScaler because the dataset may contain outliers and features with varying distributions.
StandardScaler transforms the data to have a mean of 0 and standard deviation of 1, which works well for models like Linear Regression. 
MinMaxScaler scales values between 0 and 1, but it is more sensitive to outliers because extreme values can compress the remaining
data into a small range.'''
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(x_train)
X_test_scaled = scaler.transform(x_test)

In [319]:
'''Implemented Linear Regression model again with scaled dataset and trained it again 
to learn new relationships between input features and flight price and compare it with without scaling data accuracy.'''
lr = LinearRegression()

lr.fit(X_train_scaled, y_train)

LinearRegression()

In [320]:
#predict the new value using scaled dataset
y_pred_scaled = lr.predict(X_test_scaled)

In [321]:
#Evaluate it with new scaled data again
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

print("MSE:", mean_squared_error(y_test, y_pred_scaled))
print("MAE:", mean_absolute_error(y_test, y_pred_scaled))
print("R2:", r2_score(y_test, y_pred_scaled))

MSE: 8304569.255930369
MAE: 1957.1738017366783
R2: 0.6019798119194084


Scaling changed the values, but not the pattern between features and price, so model performance remained unchanged.
| Metric | Value   |
| ------ | ------- |
| MSE    | 8304569 |
| MAE    | 1957    |
| R²     | 0.60    |


DECISION TREE REGRESSION

In [322]:
#Built and trained a Decision Tree Regression model for flight fare prediction.
from sklearn.tree import DecisionTreeRegressor

dt = DecisionTreeRegressor()

dt.fit(X_train_scaled, y_train)

y_pred_dt = dt.predict(X_test_scaled)

In [323]:
#Evaluated Decision Tree model performance using MAE, MSE, and R² score metrics.
print("MSE:", mean_squared_error(y_test, y_pred_dt))
print("MAE:", mean_absolute_error(y_test, y_pred_dt))
print("R2:", r2_score(y_test, y_pred_dt))

MSE: 6512468.023583904
MAE: 1422.1583850931677
R2: 0.6878713792693503


Compared with Linear regression, 
1.Average prediction error reduced from:
₹1957 -> ₹1409
2.Large prediction errors reduced significantly.
 8304569 -> 6133988
 Decision Tree handled complex patterns better
3.R2 Improved
 0.60 -> 0.706
Decision Tree Results
 | Metric | Value   |
| ------ | ------- |
| MSE    | 6133988 |
| MAE    | 1409    |
| R²     | 0.706   |


RANDOM FOREST REGRESSION

In [324]:
#Built and trained a Decision Tree Regression model for flight fare prediction.
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor()

rf.fit(X_train_scaled, y_train)

y_pred_rf = rf.predict(X_test_scaled)

In [325]:
#Evaluated Random Forest model performance using MAE, MSE, and R² score metrics.
print("MSE:", mean_squared_error(y_test, y_pred_rf))
print("MAE:", mean_absolute_error(y_test, y_pred_rf))
print("R2:", r2_score(y_test, y_pred_rf))

MSE: 3999966.5754488134
MAE: 1202.87305349085
R2: 0.8082901834385562


1.Average prediction error is around:
₹1203
Random Forest predictions are much closer to actual flight prices.
2.Large prediction mistakes reduced heavily.
3.R2 Increased Greatly
 80.6% of the variance in flight prices

In [327]:
#Hyper parameter tuning to improve the model performace
'''Since Random Forest had multiple parameters and the dataset was relatively large,
RandomizedSearchCV was used to efficiently optimize the model and improve prediction accuracy'''
#import libraries
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestRegressor

In [329]:
#model creation
rfh = RandomForestRegressor()

In [331]:
#define parameters
params = {
    'n_estimators': [100, 200, 300],
    'max_depth': [10, 20, 30, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2']
}


In [333]:
#apply randomizedsearch
random_search = RandomizedSearchCV(
    estimator=rfh,
    param_distributions=params,
    n_iter=10,
    cv=3,
    verbose=2,
    n_jobs=-1,
    random_state=42
)

In [334]:
#train tuned model
random_search.fit(X_train_scaled, y_train)

Fitting 3 folds for each of 10 candidates, totalling 30 fits


RandomizedSearchCV(cv=3, estimator=RandomForestRegressor(), n_jobs=-1,
                   param_distributions={'max_depth': [10, 20, 30, None],
                                        'max_features': ['sqrt', 'log2'],
                                        'min_samples_leaf': [1, 2, 4],
                                        'min_samples_split': [2, 5, 10],
                                        'n_estimators': [100, 200, 300]},
                   random_state=42, verbose=2)

In [336]:
#best paarmeter combination
print(random_search.best_params_)

{'n_estimators': 100, 'min_samples_split': 5, 'min_samples_leaf': 1, 'max_features': 'sqrt', 'max_depth': 30}


In [338]:
#best model
best_rf = random_search.best_estimator_

In [342]:
#predict 
y_pred_tuned = best_rf.predict(X_test_scaled)

In [343]:
#evaluation
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

print("MSE:", mean_squared_error(y_test, y_pred_tuned))
print("MAE:", mean_absolute_error(y_test, y_pred_tuned))
print("R2:", r2_score(y_test, y_pred_tuned))

MSE: 4091919.7836828846
MAE: 1265.7462888340492
R2: 0.8038830634413567


EXPRIMENTAL ANALYSIS

In [348]:
#create new sample data
new_data = {
    'Airline': 'IndiGo',
    'Source': 'Delhi',
    'Destination': 'Cochin',
    'Total_Stops': 1,
    'day': 15,
    'month': 6,
    'year':2026,
    'Dep_hours': 10,
    'Dep_minutes': 30,
    'Arrival_hour': 14,
    'Arrival_min': 20,
    'dur_hour': 3,
    'dur_min': 50,
    'Duration_total_mins':230
}


In [349]:
#convert into dataframe
import pandas as pd
new_df = pd.DataFrame([new_data])

In [354]:
'''One-Hot Encoding was applied to Source and Destination because they are categorical features without any ordinal relationship. 
It converts categories into binary (0 and 1) columns, making them suitable for machine learning models'''
new_df = pd.get_dummies(new_df)

In [357]:
x_train.columns

Index(['Total_Stops', 'day', 'month', 'year', 'Dept_hours', 'Dept_minutes',
       'Arrival_hour', 'Arrival_min', 'Airline_Air Asia', 'Airline_Air India',
       'Airline_GoAir', 'Airline_IndiGo', 'Airline_Jet Airways',
       'Airline_Jet Airways Business', 'Airline_Multiple carriers',
       'Airline_Multiple carriers Premium economy', 'Airline_SpiceJet',
       'Airline_Trujet', 'Airline_Vistara', 'Airline_Vistara Premium economy',
       'Source_Banglore', 'Source_Chennai', 'Source_Delhi', 'Source_Kolkata',
       'Source_Mumbai', 'Destination_Banglore', 'Destination_Cochin',
       'Destination_Delhi', 'Destination_Hyderabad', 'Destination_Kolkata',
       'Destination_New Delhi', 'dur_hour', 'dur_min', 'Duration_total_mins'],
      dtype='object')

In [359]:
#add missing columns
new_df = new_df.reindex(columns=x_train.columns, fill_value=0)

In [361]:
#predict
prediction = rf.predict(new_df)

C:\Users\PRABHITH VK\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2732: UserWarning: X has feature names, but RandomForestRegressor was fitted without feature names
  warnings.warn(


In [362]:
print("Predicted Flight Fare:", prediction[0])

Predicted Flight Fare: 6299.6


In [372]:
#actual fare
predictions = rf.predict(X_test_scaled)

for i in range(10):

    print("Actual:", y_test.iloc[i])

    print("Predicted:", predictions[i])

    print("----------------")

Actual: 15894
Predicted: 13390.12
----------------
Actual: 8366
Predicted: 12383.665
----------------
Actual: 3943
Predicted: 4003.79
----------------
Actual: 8085
Predicted: 8334.53
----------------
Actual: 4049
Predicted: 5889.416666666666
----------------
Actual: 4668
Predicted: 4803.82
----------------
Actual: 11411
Predicted: 12604.45
----------------
Actual: 7594
Predicted: 11898.935833333333
----------------
Actual: 7934
Predicted: 7163.38
----------------
Actual: 12692
Predicted: 11237.231666666667
----------------


Since flight pricing is dynamic and influenced by multiple external factors, the model predicts approximate fares rather than exact prices. Many predictions were close to actual values, while a few outlier samples showed larger deviations.

# Model comparison Result– Flight Fare Prediction

The objective of this project was to build a machine learning model capable of predicting flight ticket prices based on various features such as airline, source, destination, total stops, duration, and journey timing.

---

# Models Implemented
The following regression models were implemented and evaluated:

1. Linear Regression(with and without scaling)
2. Decision Tree Regressor
3. Random Forest Regressor
4. Tuned Random Forest Regressor

---

# Evaluation Metrics

The models were evaluated using the following regression metrics:

* Mean Squared Error (MSE)
* Mean Absolute Error (MAE)
* R² Score

These metrics help measure prediction accuracy and overall model performance.

---

# Model Performance Comparison

| Model                   | MSE          | MAE          | R² Score |
| ----------------------- | ------------ | ------------ | -------- |
| Linear Regression       | Higher error | Higher error | ~0.60    |
| Decision Tree Regressor | 6133988      | 1409         | 0.706    |
| Random Forest Regressor | 4028216      | 1203         | 0.806    |
| Tuned Random Forest     | 4091919      | 1265         | 0.803    |

---

# Analysis

## Linear Regression

Linear Regression showed lower performance because flight fare data contains nonlinear relationships and complex pricing patterns which cannot be captured effectively using a simple linear model.

## Decision Tree Regressor

Decision Tree improved performance compared to Linear Regression by learning nonlinear patterns. However, it showed signs of overfitting and produced less stable predictions.

## Random Forest Regressor

Random Forest Regressor achieved the best overall performance with:

* highest R² score
* lowest MAE
* better generalization capability

The ensemble learning approach helped reduce overfitting and improved prediction stability.

## Tuned Random Forest

Hyperparameter tuning was performed using RandomizedSearchCV. The tuned model produced performance very similar to the baseline Random Forest model, indicating that the default parameters were already near optimal for this dataset.

---

# Best Model for Production

Based on overall evaluation metrics and prediction stability, the Random Forest Regressor was selected as the best model for production deployment.

Reasons:

* Highest prediction accuracy
* Better handling of nonlinear relationships
* Lower prediction error
* Better generalization on unseen data

---

# Real-World Observation

While testing the model on new unseen samples, many predictions were highly accurate while a few outlier samples showed larger deviations. This is expected in real-world flight fare prediction because ticket prices are influenced by dynamic factors such as demand, seasonality, airline pricing strategy, and booking timing.

---

# Conclusion

The Random Forest Regressor successfully learned important pricing patterns from the dataset and achieved strong overall predictive performance with an R² score of approximately 0.80.

This project provided hands-on experience in:

* data preprocessing
* feature engineering
* regression modeling
* model evaluation
* hyperparameter tuning
* real-world prediction analysis

The final model is suitable for building a flight fare prediction application or deployment system.


# Report on Challenges Faced – Flight Fare Prediction Project

## Introduction

During the development of the Flight Fare Prediction project, multiple challenges were encountered in data preprocessing, feature engineering, model building, and prediction analysis. Various machine learning techniques and preprocessing methods were applied to address these challenges and improve model performance.

---

# 1. Handling Categorical Features

## Challenge

The dataset contained multiple categorical columns such as:

* Airline
* Source
* Destination

Machine learning models cannot directly process categorical string values.

## Technique Used

One Hot Encoding using:

```python
pd.get_dummies()
```

## Reason

One Hot Encoding converts categorical values into numerical binary columns, allowing machine learning algorithms to process the data effectively without introducing ordinal relationships between categories.

---

# 2. Feature Extraction from Date and Time Columns

## Challenge

The dataset contained date and time information in raw format, which could not be directly used for model training.

## Technique Used

Feature Engineering

Extracted features such as:

* Journey Day
* Journey Month
* Departure Hour
* Departure Minute
* Arrival Hour
* Duration Hours

## Reason

Extracting meaningful numerical features from datetime columns helps the model identify important travel timing patterns affecting ticket prices.

---

# 3. Handling Nonlinear Relationships

## Challenge

Flight fare pricing patterns were highly nonlinear and complex. Linear Regression failed to capture these relationships effectively.

## Technique Used

Decision Tree Regressor and Random Forest Regressor

## Reason

Tree-based algorithms are capable of learning nonlinear relationships and interactions between features more effectively than linear models.

---

# 4. Overfitting in Decision Tree

## Challenge

Decision Tree Regressor showed overfitting behavior by learning training data patterns too specifically, resulting in unstable predictions on unseen data.

## Technique Used

Random Forest Regressor

## Reason

Random Forest uses ensemble learning by combining multiple decision trees, reducing overfitting and improving generalization performance.

---

# 5. Model Optimization

## Challenge

Selecting the best parameter values for Random Forest manually was difficult and time-consuming.

## Technique Used

RandomizedSearchCV for Hyperparameter Tuning

Parameters tuned:

* n_estimators
* max_depth
* min_samples_split
* min_samples_leaf

## Reason

RandomizedSearchCV efficiently searches multiple parameter combinations while reducing computational cost compared to GridSearchCV.

---

# 6. Prediction Variance on New Data

## Challenge

Some predictions on new unseen data showed larger differences from actual flight fares.

## Technique Used

Model Evaluation using:

* MAE
* MSE
* R² Score

## Reason

Flight prices are highly dynamic and influenced by hidden real-world factors such as:

* demand fluctuations
* airline pricing strategy
* seasonal variations
* booking timing

The evaluation metrics helped measure overall model performance rather than relying on single prediction samples.

---

# 7. Data Preprocessing Consistency

## Challenge

Ensuring that new unseen data followed the same preprocessing pipeline as training data.

## Technique Used

* One Hot Encoding
* Reindexing columns using training dataset columns

## Reason

Maintaining consistent feature structure between training and prediction data is critical for accurate model predictions.

Example:

```python
new_df = new_df.reindex(columns=X_train.columns, fill_value=0)
```

---

# Conclusion

The Flight Fare Prediction project involved multiple real-world machine learning challenges related to preprocessing, feature engineering, model selection, and prediction analysis.

Through techniques such as:

* One Hot Encoding
* Feature Engineering
* Random Forest Regression
* Hyperparameter Tuning
* Performance Evaluation

the final model achieved strong predictive performance with an R² score of approximately 0.80.

The project provided valuable practical understanding of:

* end-to-end machine learning workflow
* regression modeling
* handling real-world noisy datasets
* evaluating prediction performance on unseen data
